applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py

```diff
@@ -0,0 +1,763 @@
+    if (auxiliary_MLMC_object.current_iteration == 0): # screening phase
+        '''compute parameters by least square fit'''
+        auxiliary_MLMC_object.ComputeRatesLS()
+        '''compute Bayesian variance V^c[Y_l]'''
+        auxiliary_MLMC_object.EstimateBayesianVariance(auxiliary_MLMC_object.current_number_levels)
+    else: # MLMC phase
+        '''compute estimator MLMC mean QoI'''
+        auxiliary_MLMC_object.ComputeMeanMLMCQoI()
+        '''compute parameters by least square fit'''
+        auxiliary_MLMC_object.ComputeRatesLS()
+        '''compute Bayesian variance V^c[Y_l]'''
+        auxiliary_MLMC_object.EstimateBayesianVariance(auxiliary_MLMC_object.current_number_levels)
+        '''compute total error of MLMC simulation'''
+        auxiliary_MLMC_object.ComputeTotalErrorMLMC()
+    return auxiliary_MLMC_object.rates_error,\
+    auxiliary_MLMC_object.BayesianVariance,auxiliary_MLMC_object.mean_mlmc_QoI,\
+    auxiliary_MLMC_object.TErr,auxiliary_MLMC_object.number_samples
+
+
+class StatisticalVariable(object):
+    '''The base class for statistical variables'''
+    def __init__(self, number_levels):
+        '''constructor of the class
+        Keyword arguments:
+        self : an instance of a class
+        number_levels : number of levels
+        '''
+
+        '''values of the variable, organized per level'''
+        self.values = []
+        '''mean of the variable per each level'''
+        self.mean = []
+        '''sample variance of the variable per each level'''
+        self.sample_variance = []
+        '''second moment of the variable per each level'''
+        self.second_moment = []
+        '''bias error of the variable'''
+        self.bias_error = None
+        '''statistical error of the variable'''
+        self.statistical_error = None
+        '''type of variable: scalar or field'''
+        self.type = None
+        '''number of samples of the variable'''
+        self.number_samples = [0 for _ in range(number_levels+1)]
+
+
+    '''
+    function updating mean and second moment values and computing the sample variance
+    M_{2,n} = sum_{i=1}^{n} (x_i - mean(x)_n)^2
+    M_{2,n} = M_{2,n-1} + (x_n - mean(x)_{n-1}) * (x_n - mean(x)_{n})
+    s_n^2 = M_{2,n} / (n-1)
+    '''
+    def UpdateOnepassMeanVariance(self,level,i_sample):
+        sample = self.values[level][i_sample]
+        old_mean = self.mean[level]
+        old_M2 = self.second_moment[level]
+        number_samples_level = self.number_samples[level]
+        new_mean, new_M2, new_sample_variance, number_samples_level = UpdateOnepassMeanVarianceAux_Task(sample, old_mean, old_M2, number_samples_level)
+        self.mean[level] = new_mean
+        self.second_moment[level] = new_M2
+        self.sample_variance[level] = new_sample_variance
+        self.number_samples[level] = number_samples_level
+    '''    
+    old function doing everything serial
+    '''
+    # def UpdateOnepassMeanVariance(self,level,i_sample):
```

Comment:
Do you need this code?

Reply:
Actually no, it was for the user as reminder of what the function is actually doing, but I delete it.